text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,464 @@
+
+```typescript
+const assembly = new CloudAssembly("/path/to/cloud-assembly");
+const artifacts = assembly.artifacts;
+const version = assembly.version;
+const runtime = assembly.runtime;
+```
+
+This method will also be used by the framework, as the return value of the `app.synth()` method.
+
+> Serialization will not be part of this package, as it is only used in the framework.
+
+This new package has to be **stable** to enforce `jsii` compatibility. This is ok, it actually makes perfect
+sense for it to be stable as it represents the contract the `Cloud-Assembly` needs to adhere to.
+
+#### Step 2: Change CLI dependencies
+
+The CLI should only depend on the new `cx-protocol` package, and **not** on the `cx-api` package.
+
+This means that when users install new versions of the CLI, they also get a new version of `cx-protocol`,
+but since `cx-protocol` is backwards compatible, there is no risk of breaking API.
+
+> The CLI currently depends on parts of `cx-api` that don't strictly relate to the proposed `cx-protocol` object model.
+> Analysis of those dependencies has yet to be done. Ideally, we would be able to move that code away from `cx-api` and into the CLI itself. Perhaps even at the expense of some duplication if that code is being used in the framework as well.
+> In any case, I don't think this will be a blocker for the design.
+
+#### Step 3: Validate CLI `>=` Framework
+
+To make sure we enforce that the CLI version will always be `>=` than the framework version, we will add a
+validation in the CLI, immediately after we de-serialize the `Cloud-Assembly`:
+
+in `exec.ts`
+
+```typescript
+var assembly;
+
+if (await fs.pathExists(app) && (await fs.stat(app)).isDirectory()) {
+    debug('--app points to a cloud assembly, so we by pass synth');
+    assembly = new cxapi.CloudAssembly(app);
+} else {
+    debug('--app points to an executable, let the framework do its thing');
+    await exec();
+    assembly = new cxapi.CloudAssembly(outdir);
+}
+
+if (versionNumber() < assembly.version) {
+    throw new Error(`A newer version of the CDK CLI (>= ${assembly.version}) is necessary to interact with this app`);
+}
+```
+
+This implies that the `version` in `manifest.json` will no longer be the value of `CLOUD_ASSEMBLY_VERSION`,
+it will simply be the version of the package itself. It also means, that we can completely get rid of this constant,
+and of the separation we have in the code between `cx-protocol` version and our package versions.
+
+> See a [quirk](#Quirk---CDK-Synth) that is caused by this.
+
+To actually validate this behavior, we add an integration test:
+
+\+ `test-cli-throws-on-new-framework.sh`
+
+```bash
+
+LATEST_PUBLISHED_VERSION=$(node -p 'require("package.json").version')
+NEXT_VERSION=$(bump_minor ${LATEST_PUBLISHED_VERSION})
+
+# create the cloud assembly
+cdk synth
+
+# patch it to simulate a later version
+sed -i "s/${LATEST_PUBLISHED_VERSION}/${NEXT_VERSION}/g" cdk.out/manifest.json
```

Comment:
Feels brittle. What if the `LATEST_PUBLISHED_VERSION` happens to also be a string used for something else at a given point? What if the assembly version is written to other locations?

Reply:
Brittle is putting it mildly! There are so many things wrong with this line... 

But, code snippets in the RFC only convey intent

Reply:
Sure but they can also be interpreted as normative by implementors.